From d16c8ad4c1b0e933e4aa191c178ea2eb4ef482fe Mon Sep 17 00:00:00 2001
From: Yi Yang <yi.y.yang@intel.com>
Date: Mon, 12 Dec 2016 13:06:20 +0800
Subject: [PATCH 8/8] tunnel: set udp dst-port in tunnel metadata

From a commit by Pravin Shelar did.

Signed-off-by: Yi Yang <yi.y.yang@intel.com>
---
 ofproto/tunnel.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/ofproto/tunnel.c b/ofproto/tunnel.c
index bc92b40..c393814 100644
--- a/ofproto/tunnel.c
+++ b/ofproto/tunnel.c
@@ -429,6 +429,7 @@ tnl_port_send(const struct ofport_dpif *ofport, struct flow *flow,
     }
     flow->pkt_mark = tnl_port->match.pkt_mark;
 
+    flow->tunnel.tp_dst = cfg->dst_port;
     if (!cfg->out_key_flow) {
         flow->tunnel.tun_id = cfg->out_key;
     }
-- 
2.1.0

